### DISCUSSION TITLE
Mermaid in markdown

### DISCUSSION THREAD

**Discussion by SebastienDorgan** [zauberzeug/nicegui#277] [Ideas / Feature Requests]
Hi,
I am am really new to nicegui and I try to figure out how to make mermaid work with ui.markdown elements.
Here is what I tried:
```python
async def content() -> None:
    mermaid_header =  """\
    <style>
    .mermaid-pre {
        visibility: hidden;
    }
    </style>
    """
    ui.add_head_html(mermaid_header)
    mermaid_footer = """\
        <script type="module" defer>
        import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@9/dist/mermaid.esm.min.mjs';
        mermaid.initialize({
            securityLevel: 'loose',
            startOnLoad: true
        });
        let observer = new MutationObserver(mutations => {
            for(let mutation of mutations) {
            mutation.target.style.visibility = "visible";
            }
        });
        document.querySelectorAll("pre.mermaid-pre div.mermaid").forEach(item => {
            observer.observe(item, { 
            attributes: true, 
            attributeFilter: ['data-processed'] });
        });
        </script>
        """
    ui.add_body_html(mermaid_footer)
    ui.label("Streams")
    content = """\
    ```mermaid
    graph TD;
        A-->B;
        A-->C;
        B-->D;
        C-->D;
    ```
    """
    ui.markdown(content , extras=['fenced-code-blocks', 'mermaid'])
```
unfortunately the result is not what I expected to be:
![image](https://user-images.githubusercontent.com/12728415/213860828-768c6703-e436-49e2-b7ea-4da915dfb604.png)
Can you help me?

**Comment by falkoschindler** [comment]
Very interesting! Mermaid support would definitely be a nice feature.

But I'm also struggling to get it to work. Have you tried the example at https://github.com/trentm/python-markdown2/wiki/mermaid#full-example? I get a similar result like in your screenshot, so no rendered Mermaid graph. Is this an issue for markdown2 or am I just doing something wrong?

**Reply by SebastienDorgan** [reply]
That's exactly what I tried to do, adapt the markdow2 mermaid example to display the graph with nicegui. I checked in the source code of the page that the scripts were present in the header and the body and also that the mermaid javascript was ok. There was.
I will continue to investigate on my side and I will keep you informed.

**Reply by falkoschindler** [reply]
Yes, but even without adapting anything: only executing the original "Full example" Python code and opening "test-markdown.html" with a browser does not work for me. So the problem seems to be with markdown2.

**Reply by SebastienDorgan** [reply]
Indeed

**Reply by SebastienDorgan** [reply]
I found the bug in markdown2, they simply forgot to add the "mermaid" class to the code tag

**Reply by falkoschindler** [reply]
Ah, ok. I just opened an issue: https://github.com/trentm/python-markdown2/issues/495. Do you want to post the solution over there? (Or even provide a pull request.)

**Reply by SebastienDorgan** [reply]
Thank you for opening the issue, I will check tomorrow if I can provide a PR.
In the meantime I succeed to draw a graph by simply replacing ui.markdown by ui.html in my previous code:
```python
  content = """\
  <pre class="mermaid">
  graph TD;
      A-->B;
      A-->C;
      B-->D;
      C-->D;
  </pre>
  """
  graph = ui.html(content)
```

![image](https://user-images.githubusercontent.com/12728415/213879000-1ee4d982-307e-4fc5-abe0-2d3a732cbbd3.png)

**Reply by falkoschindler** [reply]
Awesome! That would be an easy workaround until markdown2 is fixed: replacing the first `<pre>` with `<pre class="mermaid">`. I'll look for an integration in `ui.markdown` tomorrow.

**Reply by SebastienDorgan** [reply]
While analyzing the code of the master branch of markdown2 I realized that the problem was already fixed.
The original code works using this branch.
